docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+# Source Generators
+
+
+## Summary
+
+Source generators aim to enable _compile time metaprogramming_, that is, code that can be created
+at compile time and added to the compilation. Source generators will be able to read the contents of 
+the compilation before running, as well as access any _additional files_, enabling generators to
+introspect both user C# code and generator specific files.
+
+> **Note**: This proposal is separate from the [previous generator design](generators.md)
+
+### High Level Design Goals
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additional_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
+- Run _un-ordered_, each generator will see the same input compilation.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Implementation
+
+At the simplest level source generators are an implementation of `Microsoft.CodeAnalysis.ISourceGenerator`
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface ISourceGenerator
+    {
+        void Execute(SourceGeneratorContext context);
+    }
+}
+```
+
+Generator implementations are defined in external assemblies passed to the compiler
+using the same `-analyzer:` option used for diagnostic analyzers.
+
+An assembly can contain a mix of diagnostic analyzers and source generators.
+Since generators are loaded from external assemblies, a generator cannot be used to build
+the assembly in which it is defined.
+
+`ISourceGenerator` has a single `Execute` method that is called by the host (either the IDE
+or the command-line compiler). `Execute` passes an instance of `SourceGeneratorContext` that provides access 
+to the `Compilation` and allows the generator to alter it by adding source and reporting diagnostics.
+
+```csharp
+    public struct SourceGeneratorContext
+    {
+        public Compilation Compilation { get; }
+
+        // TODO: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
```

Comment:
❔ Do we need to be able to declare diagnostic descriptors?

Reply:
Yes, good catch. This actually came up in our early discussions and I forgot to add it here. I'll add a note that we need to do so.

---
In reply to: [354522628](https://github.com/dotnet/roslyn/pull/40162#discussion_r354522628) [](ancestors = 354522628)